Build failed in Win10 + WSL2

### Discussed in https://github.com/nasa/fprime/discussions/3475

<div type='discussions-op-text'>

<sup>Originally posted by **AlesKus** April 11, 2025</sup>
Hello, I have a failed build in WSL2 (Win10Pro).
The changes which cause the error were introduces several weeks ago, but I don't see any issue reported. So, maybe I'm doing something wrong. 
What I've done:
`fprime-bootstrap clone https://github.com/AlesKus/fprime.git`    (for now, my fork is synchronized with devel branch https://github.com/nasa/fprime/commit/637e812d59fb198def4547164eaaa42fffff7a3d)
`cd fprime`
`. fprime-venv/bin/activate`
`./ci/tests/Framework.bash`
The output:
```
[INFO] FP Util in /home/alexkus/fprime/fprime running generate
[INFO] FP Util in /home/alexkus/fprime/fprime running generate --ut
[INFO] FP Util in /home/alexkus/fprime/fprime running build with 28 jobs
---------------- ERROR ----------------
Failed to run 'build' in /home/alexkus/fprime/fprime
---------------------------------------
---------------- STDOUT ---------------
....
[ 19%] Built target Os_File_Stub
---------------------------------------
---------------- STDERR ---------------
/home/alexkus/fprime/fprime/Os/Posix/File.cpp:42:54: error: comparison of integer expressions of different signedness: ‘long unsigned int’ and ‘long int’ [-Werror=sign-compare]
   42 | static_assert(std::numeric_limits<FwSizeType>::max() >= std::numeric_limits<off_t>::max(),
      |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/alexkus/fprime/fprime/Os/Posix/File.cpp:10:
/home/alexkus/fprime/fprime/Os/Posix/File.cpp: In member function ‘virtual Os::FileInterface::Status Os::Posix::File::PosixFile::size(FwSizeType&)’:
/home/alexkus/fprime/fprime/Os/Posix/File.cpp:115:36: error: comparison of integer expressions of different signedness: ‘FwSizeType’ {aka ‘long unsigned int’} and ‘long int’ [-Werror=sign-compare]
  115 |         FW_ASSERT(current_position <= std::numeric_limits<off_t>::max());
      |                   ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/alexkus/fprime/fprime/Fw/Types/Assert.hpp:7:41: note: in definition of macro ‘FW_ASSERT_FIRST_ARG’
    7 | #define FW_ASSERT_FIRST_ARG(ARG_0, ...) ARG_0
      |                                         ^~~~~
/home/alexkus/fprime/fprime/Os/Posix/File.cpp:115:9: note: in expansion of macro ‘FW_ASSERT’
  115 |         FW_ASSERT(current_position <= std::numeric_limits<off_t>::max());
      |         ^~~~~~~~~
/home/alexkus/fprime/fprime/Os/Posix/File.cpp: In member function ‘virtual Os::FileInterface::Status Os::Posix::File::PosixFile::preallocate(FwSizeType, FwSizeType)’:
/home/alexkus/fprime/fprime/Os/Posix/File.cpp:145:17: error: comparison of integer expressions of different signedness: ‘FwSizeType’ {aka ‘long unsigned int’} and ‘long int’ [-Werror=sign-compare]
  145 |     if ((length > std::numeric_limits<off_t>::max()) ||
      |          ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/alexkus/fprime/fprime/Os/Posix/File.cpp:146:17: error: comparison of integer expressions of different signedness: ‘FwSizeType’ {aka ‘long unsigned int’} and ‘long int’ [-Werror=sign-compare]
  146 |         (offset > std::numeric_limits<off_t>::max()) ||
      |          ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/alexkus/fprime/fprime/Os/Posix/File.cpp:170:31: error: comparison of integer expressions of different signedness: ‘FwSizeType’ {aka ‘long unsigned int’} and ‘long int’ [-Werror=sign-compare]
  170 |             if (file_position > std::numeric_limits<FwSignedSizeType>::max() ||
      |                 ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/alexkus/fprime/fprime/Os/Posix/File.cpp:171:27: error: comparison of integer expressions of different signedness: ‘FwSizeType’ {aka ‘long unsigned int’} and ‘long int’ [-Werror=sign-compare]
  171 |                 file_size > std::numeric_limits<FwSignedSizeType>::max()) {
      |                 ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/alexkus/fprime/fprime/Os/Posix/File.cpp: In member function ‘virtual Os::FileInterface::Status Os::Posix::File::PosixFile::read(U8*, FwSizeType&, Os::FileInterface::WaitType)’:
/home/alexkus/fprime/fprime/Os/Posix/File.cpp:234:14: error: comparison of integer expressions of different signedness: ‘FwSizeType’ {aka ‘long unsigned int’} and ‘long int’ [-Werror=sign-compare]
  234 |     if (size > std::numeric_limits<ssize_t>::max()) {
      |         ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/alexkus/fprime/fprime/Os/Posix/File.cpp: In member function ‘virtual Os::FileInterface::Status Os::Posix::File::PosixFile::write(const U8*, FwSizeType&, Os::FileInterface::WaitType)’:
/home/alexkus/fprime/fprime/Os/Posix/File.cpp:275:14: error: comparison of integer expressions of different signedness: ‘FwSizeType’ {aka ‘long unsigned int’} and ‘long int’ [-Werror=sign-compare]
  275 |     if (size > std::numeric_limits<ssize_t>::max()) {
      |         ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
make[2]: *** [F-Prime/Os/Posix/CMakeFiles/Os_File_Posix.dir/build.make:63: F-Prime/Os/Posix/CMakeFiles/Os_File_Posix.dir/File.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:6920: F-Prime/Os/Posix/CMakeFiles/Os_File_Posix.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make: *** [Makefile:84: all] Error 2
[ERROR] CMake erred with return code 2
```
Versions:
```
fprime-util version-check
Operating System: Linux
CPU Architecture: x86_64
Platform: Linux-5.10.74.3-microsoft-standard-WSL2+-x86_64-with-glibc2.29
Python version: 3.8.10
CMake version: 3.16.3
Pip version: 25.0.1
Pip packages:
    fprime-tools==3.6.1
    fprime-gds==3.6.1
    fprime-fpp-*==3.0.0a7
Project submodules:
```

It looks somewhat similar to https://github.com/nasa/fprime/issues/3030. But, again, I don't see any issues or discussions about this, therefore my questions are:
1. Does anyone observe this behavior?
2. Is my command line sequence correct?
</div>

--- a/Os/Posix/File.cpp
+++ b/Os/Posix/File.cpp
@@ -6,6 +6,7 @@
 #include <unistd.h>
 #include <cerrno>
 #include <limits>
+#include <type_traits>
 
 #include <Fw/Types/Assert.hpp>
 #include <Os/File.hpp>
@@ -28,6 +29,13 @@ namespace File {
 #define USER_FLAGS (0)
 #endif
 
+// Create constants for the max limits of the signed types
+// These constants are used for comparisons with complementary unsigned types to avoid sign-compare warning
+using UnsignedOffT = std::make_unsigned<off_t>::type;
+static const UnsignedOffT OFF_T_MAX_LIMIT = static_cast<UnsignedOffT>(std::numeric_limits<off_t>::max());
+using UnsignedSSizeT = std::make_unsigned<ssize_t>::type;
+static const UnsignedSSizeT SSIZE_T_MAX_LIMIT = static_cast<UnsignedSSizeT>(std::numeric_limits<ssize_t>::max());
+
 // Ensure size of FwSizeType is large enough to fit eh necessary range
 static_assert(sizeof(FwSignedSizeType) >= sizeof(off_t),
               "FwSignedSizeType is not large enough to store values of type off_t");
@@ -39,7 +47,7 @@ static_assert(sizeof(FwSizeType) >= sizeof(size_t),
 // Now check ranges of FwSizeType
 static_assert(std::numeric_limits<FwSignedSizeType>::max() >= std::numeric_limits<off_t>::max(),
               "Maximum value of FwSignedSizeType less than the maximum value of off_t. Configure a larger type.");
-static_assert(std::numeric_limits<FwSizeType>::max() >= std::numeric_limits<off_t>::max(),
+static_assert(std::numeric_limits<FwSizeType>::max() >= OFF_T_MAX_LIMIT,
               "Maximum value of FwSizeType less than the maximum value of off_t. Configure a larger type.");
 static_assert(std::numeric_limits<FwSignedSizeType>::max() >= std::numeric_limits<ssize_t>::max(),
               "Maximum value of FwSignedSizeType less than the maximum value of ssize_t. Configure a larger type.");
@@ -112,7 +120,7 @@ PosixFile::Status PosixFile::size(FwSizeType& size_result) {
     size_result = 0;
     if (Os::File::Status::OP_OK == status) {
         // Must be a coding error if current_position is larger than off_t max in Posix File
-        FW_ASSERT(current_position <= std::numeric_limits<off_t>::max());
+        FW_ASSERT(current_position <= OFF_T_MAX_LIMIT);
         // Seek to the end of the file to determine size
         off_t end_of_file = ::lseek(this->m_handle.m_file_descriptor, 0, SEEK_END);
         if (PosixFileHandle::ERROR_RETURN_VALUE == end_of_file) {
@@ -142,8 +150,8 @@ PosixFile::Status PosixFile::position(FwSizeType& position_result) {
 PosixFile::Status PosixFile::preallocate(FwSizeType offset, FwSizeType length) {
     PosixFile::Status status = Os::File::Status::NOT_SUPPORTED;
     // Check for larger size than posix supports
-    if ((length > std::numeric_limits<off_t>::max()) ||
-        (offset > std::numeric_limits<off_t>::max()) ||
+    if ((length > OFF_T_MAX_LIMIT) ||
+        (offset > OFF_T_MAX_LIMIT) ||
         (std::numeric_limits<off_t>::max() - length) < offset) {
         status = Os::File::Status::BAD_SIZE;
     }
@@ -167,8 +175,8 @@ PosixFile::Status PosixFile::preallocate(FwSizeType offset, FwSizeType length) {
             FwSizeType file_position = 0;
             status = this->position(file_position);
             // Check for overflow in seek calls
-            if (file_position > std::numeric_limits<FwSignedSizeType>::max() ||
-                file_size > std::numeric_limits<FwSignedSizeType>::max()) {
+            if (file_position > static_cast<FwSizeType>(std::numeric_limits<FwSignedSizeType>::max()) ||
+                file_size > static_cast<FwSizeType>(std::numeric_limits<FwSignedSizeType>::max())) {
                 status = Os::File::Status::BAD_SIZE;
             }
             // Only allocate when the file is smaller than the allocation
@@ -231,7 +239,7 @@ PosixFile::Status PosixFile::read(U8* buffer, FwSizeType& size, PosixFile::WaitT
                                          : size * 2;
     // POSIX APIs are implementation dependent when dealing with sizes larger than the signed return value
     // thus we ensure a clear decision: BAD_SIZE
-    if (size > std::numeric_limits<ssize_t>::max()) {
+    if (size > SSIZE_T_MAX_LIMIT) {
         return BAD_SIZE;
     }
 
@@ -272,7 +280,7 @@ PosixFile::Status PosixFile::write(const U8* buffer, FwSizeType& size, PosixFile
                                          : size * 2;
     // POSIX APIs are implementation dependent when dealing with sizes larger than the signed return value
     // thus we ensure a clear decision: BAD_SIZE
-    if (size > std::numeric_limits<ssize_t>::max()) {
+    if (size > SSIZE_T_MAX_LIMIT) {
         return BAD_SIZE;
     }
 

--- a/Svc/ComQueue/ComQueue.cpp
+++ b/Svc/ComQueue/ComQueue.cpp
@@ -7,6 +7,7 @@
 #include <Fw/Types/Assert.hpp>
 #include <Svc/ComQueue/ComQueue.hpp>
 #include "Fw/Types/BasicTypes.hpp"
+#include <type_traits>
 
 namespace Svc {
 
@@ -14,8 +15,11 @@ namespace Svc {
 // Construction, initialization, and destruction
 // ----------------------------------------------------------------------
 
+using FwUnsignedIndexType = std::make_unsigned<FwIndexType>::type;
+
 ComQueue ::QueueConfigurationTable ::QueueConfigurationTable() {
-    static_assert(std::numeric_limits<FwIndexType>::max() >= FW_NUM_ARRAY_ELEMENTS(this->entries), "Number of entries must fit into FwIndexType");
+    static_assert(static_cast<FwUnsignedIndexType>(std::numeric_limits<FwIndexType>::max()) >= FW_NUM_ARRAY_ELEMENTS(this->entries),
+                  "Number of entries must fit into FwIndexType");
     for (FwIndexType i = 0; i < static_cast<FwIndexType>(FW_NUM_ARRAY_ELEMENTS(this->entries)); i++) {
         this->entries[i].priority = 0;
         this->entries[i].depth = 0;

--- a/Svc/TlmPacketizer/TlmPacketizer.cpp
+++ b/Svc/TlmPacketizer/TlmPacketizer.cpp
@@ -74,8 +74,8 @@ void TlmPacketizer::setPacketList(const TlmPacketizerPacketList& packetList,
             entryToUse->ignored = false;
             entryToUse->id = id;
             // the offset into the buffer will be the current packet length
-            // the offset must fit within FwIndexType to allow for negative values
-            FW_ASSERT(packetLen <= std::numeric_limits<FwSignedSizeType>::max(), static_cast<FwAssertArgType>(packetLen));
+            // the offset must fit within FwSignedSizeType to allow for negative values
+            FW_ASSERT(packetLen <= static_cast<FwSizeType>(std::numeric_limits<FwSignedSizeType>::max()), static_cast<FwAssertArgType>(packetLen));
             entryToUse->packetOffset[pktEntry] = static_cast<FwSignedSizeType>(packetLen);
 
             packetLen += packetList.list[pktEntry]->list[tlmEntry].size;

